frontend_server/lib/server.dart

```diff
@@ -120,8 +124,11 @@ class _FrontendCompiler implements CompilerInterface {
     final String boundaryKey = new Uuid().generateV4();
     _outputStream.writeln("result $boundaryKey");
     final Uri sdkRoot = _ensureFolderPath(options['sdk-root']);
+    final String byteStorePath = options['byte-store'];
     final CompilerOptions compilerOptions = new CompilerOptions()
-      ..byteStore = new MemoryByteStore()
+      ..byteStore = byteStorePath != null
+          ? new FileByteStore(byteStorePath)
```

Comment:
`FileByteStore` is unlimited, it does not evict older unused keys. Applications that manage the cache (e.g. when you know that you're the only user of the cache) should use `EvictingFileByteStore` with a reasonable maximum cache size.

Reply:
`ByteStore` uses `String` keys to access `List<byte>` values.
In case of `FileStore` these keys are practically file names.

When `EvictingFileByteStore` sees that the cache size might be too big, it removes files with oldest last access time. This procedure runs once at start, and after every `1/8` of the maximum size written.

If there are results for keys that `KernelDriver` wants to read, it will use them. Otherwise it will compute them (parse, compile, etc) and put into `ByteStore`. IKG is a wrapper around `KernelDriver`.

Reply:
So it sounds like unless user renames whole bunch of files, the amount of unused keys will be small.

On other hand if size for `EvictingFileByteStore` won't accommodate complete application, operations like `restart`(that need produce kernel binary for the complete application) would be very slow as they will be recompiling evicted pieces.

Does that sound right?

Reply:
Ah, keys are "file names", but not "input file names" :-)

Actually they are MD5 hashes of some data - content of files, API of files, combination of API hashes, all this salted with `KernelDriver` version, options and SDK. We cache kernel files for each library cache with a key that is MD5 of the library files content, plus hashes of APIs of the transitive closure of all files referenced by the libraries.

Change one file, and this might cause different hashes, i.e. keys, for many libraries, so we will recompile them, and we will write them into `ByteStore`. Or maybe the change is in a method body, so just one library should be recomputed.

When you undo your change, or switch git branch, it is possible that there are still older results available, so `KernelDriver` will able to serve them from the `ByteStore`.

`restart` usually does not actually recompute anything, it just sends kernels for all libraries to the client. These kernels are most probably still cached. I can think out cases when they will not be, but probably mostly in artificial situations.

Reply:
> Actually they are MD5 hashes of some data - content of files, API of files, ...

Got you. In that case with every hot reload the cache will keep growing, which is not ideal.

>`restart` usually does not actually recompute anything, it just sends kernels for all libraries to the client. These kernels are most probably still cached. I can think out cases when they will not be, but probably mostly in artificial situations.

They are still cached unless they were evicted due to file store size constraint, right? Could it happen that for large application not all kernel files will fit into `EvictingFileByteStore`?

It seems that rather than having fixed size constraint, better way to define life cycle of cache would be in terms of time unused items stay on disk before being deleted. I would imagine deleting one-day old items, for example, would be quite acceptable regardless of how much space the cache actually occupies on disk.

Reply:
> Got you. In that case with every hot reload the cache will keep growing, which is not ideal.

Yes, the cache will grow, but its size will not be (much) greater than the maximum size.

> They are still cached unless they were evicted due to file store size constraint, right? Could it happen that for large application not all kernel files will fit into EvictingFileByteStore?

Yes, this is possible.
In this case these results will be recomputed.

> It seems that rather than having fixed size constraint, better way to define life cycle of cache would be in terms of time unused items stay on disk before being deleted. I would imagine deleting one-day old items, for example, would be quite acceptable regardless of how much space the cache actually occupies on disk.

I like the idea of limiting TTL, but without limit on the cache size we risk overflowing the file system if we generate new results again and again.

Reply:
I am not sure I understand the need for such a complicated caching mechanism which has all these caveats described above. If I understand correctly we need a simple 'library' to 'kernel_file mapping', once the kernel files are produced for a library and the VM 'accepts the reload' this kernel file needs to be cached for the library.

I feel we are trying to over optimize for a use case (users undoing and redoing different version of their files) that is not common. We will treat undos and redos as a new change and generate the kernel_file for the library again. That is how the current system works and performance for this path has never been an issue.